source/extensions/filters/network/common/redis/client_impl.h

```diff
@@ -94,19 +99,23 @@ class ClientImpl : public Client, public DecoderCallbacks, public Network::Conne
   };
 
   struct PendingRequest : public PoolRequest {
-    PendingRequest(ClientImpl& parent, PoolCallbacks& callbacks);
+    PendingRequest(ClientImpl& parent, PoolCallbacks& callbacks, std::string command);
     ~PendingRequest() override;
 
     // PoolRequest
     void cancel() override;
 
     ClientImpl& parent_;
     PoolCallbacks& callbacks_;
+    std::string command_;
```

Comment:
Consider using your StatNameSet to store command_ as a StatName only. Then when you pass them around you don't need to do a bunch of string-copies, and it will enable you to compose all the stat-names strictly from other stat-names via SymbolTable::join(), which should be faster and lock-free.

If you know the most likely set of Redis commands you can remember them as builtins in your StatNameSet at construction time, and never thereafter have to take locks to get StatNames from known commands.

Reply:
Per my other comment on how commands are extracted from Redis `RespValue`, I don't think we have a great way to do this. We'll always need to map a string command to a `StatName`.

Reply:
Yes we will need one map lookup for the command since it is received in the wire as a string.

But ideally you should have a finite set of known commands that you can register as built-ins at construction time so that the StatNameSet can finish the lookup in its built-in map, which doesn't require a lock.

If you receive an unexpected command then the StatNameSet will fall back to taking a lock and doing the lookup in its dynamic map, which then falls back to the global symbol table.

Reply:
Ok, I've updated this and will push update here shortly.